site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/repository/HapiFhirRepository.java

```diff
@@ -0,0 +1,110 @@
+import org.springframework.stereotype.Component;
+import ca.uhn.fhir.rest.server.exceptions.BaseServerResponseException;
+import uk.ac.ox.ndph.mts.site_service.exception.RestException;
+
+/**
+ * Implement FhirRepository interface using HAPI client sdk and backed up by
+ * a FHIR store.
+ */
+@Component
+public class HapiFhirRepository implements FhirRepository {
+
+    private static final String ORGANIZATION_ENTITY_NAME = "Organization";
+    private static final String RESEARCHSTUDY_ENTITY_NAME = "ResearchStudy";
+
+    private final FhirContextWrapper fhirContextWrapper;
+    private final Logger logger = LoggerFactory.getLogger(HapiFhirRepository.class);
+
+    @Value("${fhir.uri}")
+    private String fhirUri = "";
+
+    HapiFhirRepository(FhirContextWrapper fhirContextWrapper) {
+        this.fhirContextWrapper = fhirContextWrapper;
+    }
+
+    /**
+     * @param organization the organization to save.
+     * @return id of the saved organization
+     */
+    public String saveOrganization(Organization organization) {
+        // Log the request
+        logger.info(FhirRepo.SAVE_REQUEST.message(),
+                fhirContextWrapper.prettyPrint(organization));
+
+        Bundle responseBundle;
+        try {
+            responseBundle = fhirContextWrapper.executeTrasaction(fhirUri, 
+                bundle(organization, ORGANIZATION_ENTITY_NAME));
+        } catch (BaseServerResponseException e) {
+            logger.warn(FhirRepo.UPDATE_ERROR.message(), e);
+            throw new RestException(e.getMessage(), e);
+        }
+        IBaseResource responseElement = extractResponseResource(responseBundle);
+
+        // Log the response
+        logger.info(FhirRepo.SAVE_RESPONSE.message(),
+                    fhirContextWrapper.prettyPrint(responseElement));
+
+        return responseElement.getIdElement().getIdPart();
+    }
+
+    /**
+     * @param researchStudy the researchStudy to save.
+     * @return ResearchStudy
+     */
+    public String saveResearchStudy(ResearchStudy researchStudy) {
+        // Log the request
+
+        logger.info(FhirRepo.SAVE_REQUEST.message(),
+                    fhirContextWrapper.prettyPrint(researchStudy));
+
+        Bundle responseBundle;
+        try {
+            responseBundle = fhirContextWrapper.executeTrasaction(fhirUri,
+                    bundle(researchStudy, RESEARCHSTUDY_ENTITY_NAME));
+        } catch (BaseServerResponseException e) {
+            logger.warn(FhirRepo.UPDATE_ERROR.message(), e);
+            throw new RestException(e.getMessage(), e);
+        }
+        IBaseResource responseElement = extractResponseResource(responseBundle);
+
+        // Log the response
+        logger.info(FhirRepo.SAVE_RESPONSE.message(),
+                    fhirContextWrapper.prettyPrint(responseElement));
+
+        return responseElement.getIdElement().getIdPart();
+    }
+
+    private IBaseResource extractResponseResource(Bundle bundle) throws RestException {
```

Comment:
why is this named extractResourceBundle? the purpose is to check size and throw an error? rename to validateSize or similar?

Reply:
Taking forward as tech debt.